pkg/controller/expcertificates/issuing/issuing_controller.go

```diff
@@ -200,22 +229,46 @@ func (c *controller) ProcessItem(ctx context.Context, key string) error {
 		return nil
 	}
 
-	switch cond.Reason {
 	// If the certificate request has failed, set the last failure time to now,
 	// and set the Issuing status condition to False with reason.
-	case cmapi.CertificateRequestReasonFailed:
+	if cond.Reason == cmapi.CertificateRequestReasonFailed {
 		return c.failIssueCertificate(ctx, log, crt, req)
+	}
 
-		// If the CertificateRequest is valid, verify its status and update
-		// accordingly.
-	case cmapi.CertificateRequestReasonIssued:
-		return c.issueCertificate(ctx, log, nextRevision, crt, req)
+	// Verify the CSR options match what is requested in certificate.spec.
+	// If there are violations in the spec, then the requestmanager will handle this.
+	requestViolations, err := certificates.RequestMatchesSpec(req, crt.Spec)
+	if err != nil {
+		return err
+	}
+	if len(requestViolations) > 0 {
+		log.Info("CertificateRequest does not match Certificate")
+		return nil
+	}
 
-	// CertificateRequest is not in a final state so do nothing.
-	default:
-		log.V(4).Info("CertificateRequest not in final state, waiting...")
+	// If public key does not match, do nothing (keymanager will handle this).
+	csr, err := utilpki.DecodeX509CertificateRequestBytes(req.Spec.CSRPEM)
+	if err != nil {
+		return err
+	}
+	publicKeyMatchesCSR, err := utilpki.PublicKeyMatchesCSR(pk.Public(), csr)
+	if err != nil {
+		return err
+	}
+	if !publicKeyMatchesCSR {
+		logf.WithResource(log, nextPrivateKeySecret).Info("next private key does not match CSR public key, waiting for keymanager controller")
 		return nil
 	}
+
+	// If the CertificateRequest is valid and ready, verify its status and issue
+	// accordingly.
+	if cond.Reason == cmapi.CertificateRequestReasonIssued {
+		return c.issueCertificate(ctx, log, nextRevision, crt, req, pkData)
+	}
+
+	// CertificateRequest is not in a final state so do nothing.
+	log.V(4).Info("CertificateRequest not in final state, waiting...")
```

Comment:
what if the `Reason` is unset or unrecognised? It'd be good to log/surface that the request is in a state we don't recognise I think